### ISSUE TITLE
Generate SBOMs and Add Container/Dependency Vulnerability Scanning to CI

### ISSUE THREAD

**Issue by Inkman007** [Ledger-Lenz/Ledgerlens-core#323]
## Summary
`.github/workflows/ci.yml` currently runs two jobs — `test` (matrix over Python 3.10/3.12: `pip install`, `ruff check .`, `pytest --cov=detection --cov=ingestion --cov=config -q`) and `benchmark` (`make benchmark-check`) — with no SBOM generation and no dependency or container vulnerability scanning at any stage. `Dockerfile` builds a multi-stage, non-root container image (`ledgerlens` UID 1000, builder/runtime split) but nothing scans the resulting image or the `requirements.txt` dependency tree before or after build. This issue adds CycloneDX/Syft SBOM generation for both the Python dependency tree and the built container image, Trivy or Grype scanning with a severity-gated CI failure, and Sigstore/Cosign image signing alongside the existing ED25519 model-signing scheme.

## Background & Context
The `test` job in `.github/workflows/ci.yml` installs from `requirements.txt` (which pins `scikit-learn>=1.4.0`, `xgboost>=2.0.0`, `lightgbm>=4.3.0`, `cryptography>=42.0.0`, `py_ecc>=7.0.0`, and others) with no lockfile pinning transitive versions and no automated check for known-vulnerable versions landing in that resolution. The `Dockerfile`'s builder stage (`FROM python:3.12-slim AS builder`) installs `gcc`/`g++`/`libffi-dev` and compiles wheels into `/install`, then the runtime stage copies only `/install` and the app source into a fresh `python:3.12-slim` base — a good multi-stage hygiene pattern (build tools excluded from the final image, non-root `ledgerlens` user, `HEALTHCHECK` configured) but the resulting image is never scanned for OS-package or Python-package CVEs, and there is no SBOM artifact produced anywhere in the pipeline for downstream consumers or auditors to inspect.

This gap sits alongside — and is complementary to, not a duplicate of — the existing `detection/model_signing.py` ED25519 scheme documented in `docs/model_signing.md`, which protects the integrity/authenticity of trained `.joblib`/model artifacts specifically. That scheme has no bearing on the container image or the Python dependency supply chain; a compromised or vulnerable third-party package (e.g., a malicious `scikit-learn` release, or a known-CVE version of `cryptography` silently in the resolved tree) would not be caught by model signing at all. Given this project's security posture elsewhere — SSRF protection on webhook URLs and `EVM_PROVIDERS`, HTTPS-only enforcement, HMAC-signed webhooks, wallet-address masking in logs — the absence of any supply-chain scanning gate is a clear and addressable gap for a project positioning itself as production-grade fraud-detection infrastructure that other Stellar protocols compose against.

## Objectives
- [ ] Add a `sbom` CI job that generates a CycloneDX SBOM for the Python dependency tree via `cyclonedx-bom` (or `cyclonedx-py`) from `requirements.txt`, uploaded as a build artifact
- [ ] Add container image SBOM generation via `syft` (Anchore) against the built `Dockerfile` image, in CycloneDX or SPDX format, uploaded as a build artifact and optionally attached to GitHub Releases
- [ ] Add a `scan` CI job running Trivy (or Grype) against both the Python dependency SBOM and the built container image, failing the build on any `CRITICAL` or unpatched `HIGH` severity finding (configurable via `.trivyignore`/equivalent allowlist for accepted/false-positive findings with documented justification)
- [ ] Add Sigstore/Cosign keyless signing of the built container image in the release/publish workflow, so consumers can verify provenance via `cosign verify`
- [ ] Publish the SBOM and scan results as CI job summaries / artifacts so they're visible without downloading raw JSON
- [ ] Document a documented exception/waiver process for findings that cannot be immediately remediated (e.g., no upstream fix available), with a required expiry date on each waiver so they don't silently become permanent
- [ ] Add a `make sbom` / `make scan` local target mirroring the CI steps so contributors can reproduce failures before pushing

## Technical Requirements

**New CI jobs** (`.github/workflows/ci.yml`):
```yaml
  sbom:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - name: Generate Python SBOM (CycloneDX)
        run: |
          pip install cyclonedx-bom
          cyclonedx-py requirements -i requirements.txt -o sbom-python.cdx.json
      - name: Build image
        run: docker build -t ledgerlens-core:ci .
      - name: Generate container SBOM (Syft)
        uses: anchore/sbom-action@v0
        with:
          image: ledgerlens-core:ci
          format: cyclonedx-json
          output-file: sbom-container.cdx.json
      - uses: actions/upload-artifact@v4
        with:
          name: sboms
          path: |
            sbom-python.cdx.json
            sbom-container.cdx.json

  scan:
    needs: sbom
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - name: Build image
        run: docker build -t ledgerlens-core:ci .
      - name: Trivy scan (image)
        uses: aquasecurity/trivy-action@master
        with:
          image-ref: ledgerlens-core:ci
          severity: "CRITICAL,HIGH"
          exit-code: "1"
          trivyignores: ".trivyignore"
      - name: Trivy scan (Python deps)
        uses: aquasecurity/trivy-action@master
        with:
          scan-type: "fs"
          scan-ref: "requirements.txt"
          severity: "CRITICAL,HIGH"
          exit-code: "1"
          trivyignores: ".trivyignore"

  sign-image:
    needs: scan
    if: github.event_name == 'push' && github.ref == 'refs/heads/main'
    runs-on: ubuntu-latest
    permissions:
      id-token: write   # required for Sigstore keyless signing
      contents: read
    steps:
      - uses: actions/checkout@v4
      - uses: sigstore/cosign-installer@v3
      - name: Build and push image
        run: |
          docker build -t ghcr.io/ledger-lenz/ledgerlens-core:${{ github.sha }} .
          docker push ghcr.io/ledger-lenz/ledgerlens-core:${{ github.sha }}
      - name: Sign image (keyless)
        run: cosign sign --yes ghcr.io/ledger-lenz/ledgerlens-core:${{ github.sha }}
```

**`.trivyignore` format** (with mandatory expiry comments):
```
# CVE-YYYY-NNNNN: <package> — no upstream fix available as of YYYY-MM-DD.
# Waiver expires YYYY-MM-DD; re-evaluate on next dependency update cycle.
CVE-YYYY-NNNNN
```

## Security Considerations
- SBOM generation and scanning must run on **every** PR touching `requirements.txt`, `pyproject.toml`, or `Dockerfile`, not only on a schedule — supply-chain issues need to be caught before merge, not discovered after release
- Cosign keyless signing uses GitHub's OIDC token (`id-token: write` permission) and Sigstore's public transparency log (Rekor) — no long-lived signing key is stored as a repository secret, avoiding a new secret-management burden
- The `.trivyignore` waiver process must require an expiry date on every entry (see format above) and a follow-up issue reference, so waivers cannot silently persist past their justification
- Findings in the container base image (`python:3.12-slim`) versus LedgerLens's own dependencies should be triaged separately — base-image CVEs are remediated by bumping the base image tag, not by touching `requirements.txt`
- SBOM artifacts must not include any secrets or environment-specific values (they should not, since they only reflect package metadata, but this should be explicitly verified in the PR that introduces this)
- Signing must occur only on the `main` branch push path (post-merge), never on untrusted PR branches, to prevent a malicious PR from obtaining a validly signed image for arbitrary content

## Testing Requirements
- CI smoke test: `sbom` job produces both `sbom-python.cdx.json` and `sbom-container.cdx.json` as non-empty, schema-valid CycloneDX documents (validate with `cyclonedx-cli validate`)
- CI smoke test: `scan` job fails when a deliberately vulnerable test dependency (e.g., pinning an old `cryptography` version with a known CVE in a scratch branch) is introduced, and passes on the current clean `requirements.txt` — verify this manually once during PR review, documented in the PR description
- CI smoke test: `.trivyignore` entries without an expiry-date comment are rejected by a small validation script (`scripts/check_trivyignore_expiry.py`) run as part of the `scan` job
- `scripts/check_trivyignore_expiry.py` unit tests (`tests/test_trivyignore_expiry.py`): expired waiver entries cause the check to fail; well-formed unexpired entries pass
- Local reproduction: `make sbom` and `make scan` targets produce equivalent output to the CI job on a contributor's machine (documented, not necessarily unit-tested)

## Documentation Requirements
- New `docs/supply_chain_security.md`: SBOM generation/consumption, scan severity policy, waiver process with expiry requirement, and Cosign image verification instructions for consumers (`cosign verify ghcr.io/ledger-lenz/ledgerlens-core:<tag>`)
- `README.md`: add a short "Supply Chain Security" note (near the Docker section) linking to the new doc and documenting how to verify a published image's signature
- `docs/model_signing.md`: add a cross-reference noting this scheme covers `.joblib`/model artifacts specifically, while `docs/supply_chain_security.md` covers the container image and Python dependency tree

## Definition of Done
- [ ] All objectives completed
- [ ] Tests pass (`pytest`)
- [ ] No regressions on existing test suite
- [ ] PR reviewed and approved

## For Contributors
**When applying for this issue, please specify:**
- Your area of specialty
- Relevant experience with: CycloneDX/Syft SBOM tooling, Trivy or Grype vulnerability scanning, Sigstore/Cosign keyless signing, and GitHub Actions workflow design
- Your approach or initial thoughts on the implementation
- Estimated time to complete

**Ideal contributor profile:** A DevSecOps/supply-chain-security engineer with production experience wiring SBOM generation and vulnerability scanning into CI/CD pipelines, familiar with Sigstore's keyless signing model and comfortable setting a sane severity-gating policy that doesn't create constant noise.

**Comment by Inkman007** [comment]
Closing as part of a full issue-tracker reset: the project's open issues have been replaced with a new curated set of 125 advanced-level issues covering the codebase end to end. No action needed here; if this item is still relevant, please open a fresh issue referencing it.
